dspy/functional/typed_predictor_signature.py

```diff
@@ -0,0 +1,130 @@
+        """
+        Return a DSPy Signature class that can be used to extract the output parameters.
+
+        :param pydantic_class_for_dspy_input_fields: Pydantic class that defines the DSPy InputField's.
+        :param pydantic_class_for_dspy_output_fields: Pydantic class that defines the DSPy OutputField's.
+        :param prefix_instructions: Optional text that is prefixed to the instructions.
+        :return: A DSPy Signature class optimizedfor use with a TypedPredictor to extract structured information.
+        """
+        if prefix_instructions:
+            prefix_instructions += "\n\n"
+        instructions = prefix_instructions + "Use only the available information to extract the output fields.\n\n"
+        dspy_fields = {}
+        for field_name, field in pydantic_class_for_dspy_input_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:
+                field.validate_default = False
+
+            input_field = InputField(desc=field.description)
+            dspy_fields[field_name] = (field.annotation, input_field)
+
+        for field_name, field in pydantic_class_for_dspy_output_fields.model_fields.items():
+            if field.default and 'typing.Annotated' in str(field.default):
+                raise ValueError(f"Field '{field_name}' is annotated incorrectly. See 'Constraints on compound types' in https://docs.pydantic.dev/latest/concepts/fields/")
+
+            is_default_value_specified, is_marked_as_optional, inner_field = cls._process_field(field)
+            if is_marked_as_optional:
+                if field.default is None or field.default is PydanticUndefined:
+                    field.default = 'null'
+                field.description = inner_field.description
+                field.examples = inner_field.examples
+                field.metadata = inner_field.metadata
+                field.json_schema_extra = inner_field.json_schema_extra
+            else:   
+                field.validate_default = False
+
+            if field.default is PydanticUndefined:
+                raise ValueError(
+                    f"Field '{field_name}' has no default value. Required fields must have a default value. "
```

Comment:
ah interesting concept here, requiring a default value for every required output field, why?

Reply:
Users may provide input without passing the value of a required parameter. So how to indicate that the required value was missing in the input, without throwing an exception? And without forcing the LM to hallucinate a required missing value?

Luckily, you can specify that Pydantic default value should not be subject to validation. This means that regardless of the type (str, int, float, ...), you can specify a default value of "NOT_FOUND" and this signature will correctly detect and return it without hallucinating. 

No wrap validators necessary, which in any case don't distinguish between invalid and missing

Reply:
The unique nature of agentic applications is that we cannot assume user input is being validated in the UI for required input. Combine this with LM ability to hallucinate when forced to do so...